src/ddscxx/src/org/eclipse/cyclonedds/pub/AnyDataWriterDelegate.cpp

```diff
@@ -116,6 +121,29 @@ AnyDataWriterDelegate::write_cdr(
         data->payload().size() + 4);
 
     ser_data->statusinfo = statusinfo;
+
+#ifdef DDSCXX_HAS_SHM
+    // if shared memory is supported by the writer
+    if(dds_is_shared_memory_available(writer)) {
+        void *iox_chunk ;
+        // request a loan from the shared memory buffer
+        ret = dds_loan_shared_memory_buffer(writer,
+                                            data->payload().size() + data->encoding().size(),  // include the size for the encoding (CDR header)
+                                            &iox_chunk);
+        ISOCPP_DDSC_RESULT_CHECK_AND_THROW(ret, "write_cdr - Loaning of chunk failed");
+        ISOCPP_BOOL_CHECK_AND_THROW(iox_chunk, ISOCPP_NULL_REFERENCE_ERROR, "write_cdr - Loaning of chunk failed");
+        // copy the header
+        memcpy(iox_chunk, data->encoding().data(), data->encoding().size());
+        // copy the actual data
+        memcpy(static_cast<unsigned char *>(iox_chunk) + data->encoding().size(),
```

Comment:
Copying the serialized data later after it has been serialized is also a source of severe inefficiencies. Ideally it should already have been serialized to the buffer. I assume that due to the existing architecture it may be needed n other memory of the serialized sample representation as well, but I am not sure.

Reply:
Since we don't have a user API to loan the memory and fill the serialized sample, we get the serialized sample as `CDRBlob` then copy this into the `iox_chunk`, which will then be eventually transferred using shm